utils/clflags.ml

```diff
@@ -577,3 +577,26 @@ let create_usage_msg program =
 
 let print_arguments program =
   Arg.usage !arg_spec (create_usage_msg program)
+
+let auto_include_otherlibs alert uncap fn =
+  if !no_std_include then raise Not_found
+  else
+    let expand = Misc.expand_directory Config.standard_library in
+    let read_lib lib = lazy (Load_path.Dir.create (expand ("+" ^ lib))) in
+    let otherlibs =
+      List.map (fun lib -> (lib, read_lib lib)) ["dynlink"; "str"; "unix"] in
+    let scan (lib, dir) =
+      let dir = Lazy.force dir in
+      let check base =
+        if uncap base = fn then begin
+          alert lib;
+          Load_path.append_dir dir;
+          Some (Filename.concat (Load_path.Dir.path dir) base)
+        end else
+          None
+      in
+      List.find_map check (Load_path.Dir.files dir)
+    in
+    match List.find_map scan otherlibs with
+    | Some base -> base
+    | None -> raise Not_found
```

Comment:
I think this works, although I think something like this would be clearer:

In `load_path.ml`
```ocaml
with Not_found ->
  (* call auto_include_callback like this instead, or the uncap version for the first call *)
  auto_include_callback (fun dir -> Misc.find_in_path_uncap [ dir ] fn)
```

here:
```ocaml
let auto_include alert find_in_dir =
  if !no_std_include then raise Not_found
  else
    List.find_map [ "dynlink"; "str"; "unix" ] ~f:(fun lib ->
      let dir = Filename.concat Config.standard_library lib in
      match find_in_dir dir with
      | Not_found -> None
      | x -> alert lib; Load_path.append_dir dir; Some x)
    (* then turn the option back into Not_found *)
```

I wonder if there's a cost to checking for file existence for every missing file. It might seem like a missing file will result in a compilation failure immediately but that's not the case: the compiler can look for cmi or cmx files that may well be missing and proceed. Not sure if that's a real concern, to be clear, but that's why I was suggesting before to special case the cmis that are known to live in these libs: it means no extra filesystem interaction.

Reply:
I just pushed an amended version - instead of using `Misc.find_in_path`, I added `Load_path.Dir.find` which allows the two call-sites in `Load_path` to pass either `Load_path.Dir.find` or `Load_path.Dir.find_uncap`. I agree that the result looks much clearer - and it still gets to keep the lazy scanning as well (because the `Dir.t` is generated lazily).